### PR TITLE
Handle embedded forms

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Handle embedded forms from the proxy

#### Ticket Link
None

#### Related Pull Requests
Redux: https://github.com/mattermost/mattermost-redux/pull/1262
Webapp: https://github.com/mattermost/mattermost-webapp/pull/6804

### CODE REVIEW

File: server/http/helloapp/embedded_wish.go
Comment (mickmister) [inline server/http/helloapp/embedded_wish.go:82]: It looks like this function is not called. Is this here to toggle between the two dialogs by editing the source here?

File: server/http/restapi/embedded_form.go
Comment (mickmister) [inline server/http/restapi/embedded_form.go:49]: We can get rid of the first block here as well

File: server/http/restapi/embedded_form.go
Comment (mickmister) [inline server/http/restapi/embedded_form.go:90]: Do we want to return an error code in the response here, or does the response need to be 200?

File: server/http/restapi/embedded_form.go
Comment (mickmister) [inline server/http/restapi/embedded_form.go:93]:
```suggestion
			if updatedPost, parseErr := postFromInterface(resp.Data["post"]); parseErr == nil {
```

File: server/http/restapi/embedded_form.go
Comment (mickmister) [inline server/http/restapi/embedded_form.go:132]: Can you link to the source of this comment here? https://github.com/mattermost/mattermost-server/blob/20491c2585475c2218f964e0a882c65deac570a5/app/integration_action.go#L57

**Review by mickmister — APPROVED** [review]
Awesome work, just some non-blocking questions/suggestions on cleaning up code

File: server/http/helloapp/embedded_wish.go
Comment (larkox) [reply server/http/helloapp/embedded_wish.go:82]: Exactly. It is in case you want to trigger a fuller example.

File: server/http/restapi/embedded_form.go
Comment (larkox) [reply server/http/restapi/embedded_form.go:90]: I believe this is expected to work with a 200 response, not an error response. The thinking behind is "the transaction was succesful, and you have to show an error message".

**Review by mickmister — APPROVED** [review]
LGTM great work :+1:

File: server/http/helloapp/embedded_wish.go
Comment (levb) [inline server/http/helloapp/embedded_wish.go:19]: constants.AppID

Comment (levb) [reply server/http/helloapp/embedded_wish.go:19]: A hypothetical question, can we envision multiple embedded forms added to a post? Should AppID be embedded in the Form?

Either way I don't like having `AppID` as a top-level prop, because it relates that a post "belongs to" an App, which should not be the case.

File: server/http/helloapp/embedded_wish.go
Comment (levb) [inline server/http/helloapp/embedded_wish.go:20]: we should decide on the term and be consistent. Perhaps, `"embedded_form"`? also, `constants.XXX`

File: server/http/restapi/embedded_form.go
Comment (levb) [reply server/http/restapi/embedded_form.go:21]: No need to close req.Body, ever, the http server takes care of that. In fact, it may be a counter-pattern, messing with connection management.

**Review by levb — CHANGES_REQUESTED** [review]
I am going to hold off a more detailed review until master is merged

File: server/http/helloapp/embedded_wish.go
Comment (larkox) [reply server/http/helloapp/embedded_wish.go:19]: I do not like much the idea of several plugins displaying a dialog in the same message. Also, I think the post does indeed belong to the app, since it is the one posting it as a form, and will be the one modifying it.

File: server/http/helloapp/embedded_wish.go
Comment (larkox) [reply server/http/helloapp/embedded_wish.go:20]: It is the same as an interactive dialog. So we should unify the two concepts. Here some proposals:
- Interactive Dialog - Embedded Dialog
- Modal Form - Embedded Form

**Review by levb — APPROVED** [review]
I still have a bunch of questions, but this is a big step forward.

File: server/api/call.go
Comment (larkox) [reply server/api/call.go:31]: I personally prefer a map[string]interface, since it is easier to handle. I understand that for this PR, i could have added the Post or the Error list directly as the Data, but I still feel more comfortable using map[string]interface{}. Nevertheless, I am open to revert back that change.

File: server/api/embedded.go
Comment (larkox) [reply server/api/embedded.go:4]: Since this one is strictly used on embedded forms, I wonder if it makes sense to have it separated. But I am open to unify.

File: server/api/call.go
Comment (levb) [reply server/api/call.go:31]: IMO, `Data` is to return any data the app wants to return from a successful invocation, usually some sort of a JSON result. The pointer here should be as generic as possible, thus interface{}.

For other use-cases (returning the Form, for example) I am inclined to add typed, dedicated fields. Anyway, we can decide on this later. Do a separate cleanup when we have more use-cases implemented.

File: server/api/embedded.go
Comment (levb) [reply server/api/embedded.go:4]: Gotcha. 0/5 if it walks like a duck, if it quacks like a duck, it is an AppID (aka app_id). Not an essential issue, obviously.

File: server/http/helloapp/embedded_wish.go
Comment (levb) [reply server/http/helloapp/embedded_wish.go]: 0/5 log as a separate issue to discuss/resolve later?

File: server/http/helloapp/embedded_wish.go
Comment (levb) [reply server/http/helloapp/embedded_wish.go]: Why should we ultimately support the old-style interactive dialogs? I thought it was temporary "stub", to be replaced later?

**Pull request merged**
